### ISSUE TITLE
[BUG]: Crash while reviewing

### ISSUE THREAD

**Issue by mikehardy** [ankidroid/Anki-Android#14920]
### Checked for duplicates?

- [X] This issue is not a duplicate

### What are the steps to reproduce this bug?

Not exactly sure, but I was reviewing quickly and that may have affected it
Open the deck Japanese-10k-testdeck attached here
[Japanese-10k-testdeck.apkg.gz](https://github.com/ankidroid/Anki-Android/files/13625272/Japanese-10k-testdeck.apkg.gz)
 with enough reviews set so that you can roll through a bunch of reviews quickly
Start reviewing with a mix of hard / good answers, rapidly

### Expected behaviour

Should be able to roll through the reviews, no problem

### Actual behaviour

```
12-09 12:56:49.357  3348  3852 D ReviewerServer: GET /reviewer.html
12-09 12:56:49.404  3348  3852 D ReviewerServer: GET /assets/flashcard.css
12-09 12:56:49.404  3348  3852 V ReviewerServer: OK: /assets/flashcard.css
12-09 12:56:49.404  3348  3850 D ReviewerServer: GET /assets/reviewer_extras.css
12-09 12:56:49.404  3348  3853 D ReviewerServer: GET /assets/scripts/card.js
12-09 12:56:49.404  3348  3849 D ReviewerServer: GET /assets/reviewer_extras_bundle.js
12-09 12:56:49.404  3348  3851 D ReviewerServer: GET /assets/mathjax/mathjax.css
12-09 12:56:49.405  3348  3847 D ReviewerServer: GET /assets/jquery.min.js
12-09 12:56:49.405  3348  3853 V ReviewerServer: OK: /assets/scripts/card.js
12-09 12:56:49.405  3348  3849 V ReviewerServer: OK: /assets/reviewer_extras_bundle.js
12-09 12:56:49.405  3348  3851 V ReviewerServer: OK: /assets/mathjax/mathjax.css
12-09 12:56:49.405  3348  3847 V ReviewerServer: OK: /assets/jquery.min.js
12-09 12:56:49.405  3348  3850 V ReviewerServer: OK: /assets/reviewer_extras.css
12-09 12:56:49.414  3348  3852 D ReviewerServer: GET /assets/scripts/js-api.js
12-09 12:56:49.414  3348  3852 V ReviewerServer: OK: /assets/scripts/js-api.js
12-09 12:56:49.418  3348  3852 D ReviewerServer: GET /1066.jpg
12-09 12:56:49.419  3348  3852 W ReviewerServer: not found: /1066.jpg
12-09 12:56:49.426   548   620 I MediaFocusControl: abandonAudioFocus() from uid/pid 10109/752 clientId=android.media.AudioManager@fbcf448
12-09 12:56:49.426   752  3863 V MediaPlayer: resetDrmState:  mDrmInfo=null mDrmProvisioningThread=null mPrepareDrmInProgress=false mActiveDrmScheme=false
12-09 12:56:49.426   752  3863 V MediaPlayer: cleanDrmObj: mDrmObj=null mDrmSessionId=null
12-09 12:56:49.426   459  1805 D NuPlayerDriver: reset(0xb400007948246950) at state 6
12-09 12:56:49.427   459  3871 I hw-BpHwBinder: onLastStrongRef automatically unlinking death recipients
12-09 12:56:49.427   463   893 I hw-BpHwBinder: onLastStrongRef automatically unlinking death recipients
12-09 12:56:49.427   463   893 I hw-BpHwBinder: onLastStrongRef automatically unlinking death recipients
12-09 12:56:49.428   463   893 V C2Store : in ~ComponentModule
12-09 12:56:49.428   463   893 V C2Store : unloading dll
12-09 12:56:49.429   459  3865 D NuPlayerDriver: notifyResetComplete(0xb400007948246950)
12-09 12:56:49.429   459  1805 W AMessage: failed to post message as target looper for handler 0 is gone.
12-09 12:56:49.432  3348  3849 D ReviewerServer: GET /_jlpt.js
12-09 12:56:49.432  3348  3849 W ReviewerServer: not found: /_jlpt.js
12-09 12:56:49.435  3348  3348 I chromium: [INFO:CONSOLE(338)] "Uncaught TypeError: pitch is not a function", source: http://127.0.0.1:43987/reviewer.html (338)
12-09 12:56:49.435  3348  3348 D AbstractFlashcardViewer$CardViewerWebClient: Java onPageFinished triggered: http://127.0.0.1:43987/reviewer.html#answer
12-09 12:56:49.458   548   620 W WindowManager: Attempted to add application window with unknown token Token{419eaff ActivityRecord{4dcf0cc u0 com.ichi2.anki.debug/com.ichi2.anki.pages.PagesActivity} t-1 f}}}.  Aborting.
12-09 12:56:49.458  3348  3348 D AndroidRuntime: Shutting down VM
--------- beginning of crash
12-09 12:56:49.460  3348  3348 E AndroidRuntime: FATAL EXCEPTION: main
12-09 12:56:49.460  3348  3348 E AndroidRuntime: Process: com.ichi2.anki.debug, PID: 3348
12-09 12:56:49.460  3348  3348 E AndroidRuntime: android.view.WindowManager$BadTokenException: Unable to add window -- token android.os.BinderProxy@8edce0a is not valid; is your activity running?
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.view.ViewRootImpl.setView(ViewRootImpl.java:1312)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:405)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:148)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.app.Dialog.show(Dialog.java:352)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at androidx.appcompat.app.AlertDialog$Builder.show(AlertDialog.java:1008)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at com.ichi2.anki.pages.PageChromeClient.onJsAlert(PageChromeClient.kt:63)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at mb.run(chromium-SystemWebView.apk-default-495156103:629)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:942)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:201)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:288)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7872)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
12-09 12:56:49.460  3348  3348 D UsageAnalytics: sendAnalyticsException() description/fatal: android.view.WindowManager$BadTokenException: Unable to add window -- token android.os.BinderProxy@8edce0a is not valid; is your activity running?/true
```

Looks like our web page handler tried to add an alert dialog but the page was gone already? Specific line:

> 12-09 12:56:49.460  3348  3348 E AndroidRuntime: 	at com.ichi2.anki.pages.PageChromeClient.onJsAlert(PageChromeClient.kt:63)

### Debug info

```text
built from main at commit e04d9be7cdab47044057ddd9f07ef634bb99e382 (current as of this typing)
```

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [X] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by mikehardy** [comment]
No idea why the line numbers are messed up, but here's the source link:

https://github.com/ankidroid/Anki-Android/blob/e04d9be7cdab47044057ddd9f07ef634bb99e382/AnkiDroid/src/main/java/com/ichi2/anki/pages/PageChromeClient.kt#L36-L40

Perhaps this onJsDialog and onJsConfirm methods need to have their dialog show lines in try/catch and eat exceptions log them as .w perhaps?

**Comment by mikehardy** [comment]
Had a thought - perhaps they should be converted to snackbars or toasts if they can't display? Just thinking out loud though - may be that we can't say for sure if conversion is valid for any given dialog pop or confirm

**Comment by david-allison** [comment]
We need to figure out what happens if:

1. The app is minimised and we get an alert
2. The review is complete and we can an alert

----

* There is no need to handle `JsResult` here, apparently it's unused for `alert()`
* We've produced toasts in the past where the user had minimised the app, and it wasn't a good UX
* That this may allow a someone with the ability to post toasts when the app is minimised
* We can almost certainly register the `PageChromeClient` as a `DefaultLifecycleObserver` and know whether the app is visible or not
* We can likely save one or more messages and show a dialog when the user restores the app

**Comment by BrayanDSO** [comment]
IMO, the PageChromeClient shouldn't be showing dialogs itself, nor being aware of the app lifecycle. 

Using `Flow` and subscribing the responsible activity/fragment to it would be better IMO. Or use the ActivityAgnosticDialogs stuff oakkitten added a while ago

**Comment by mikehardy** [comment]
This may be a *little* aggressive, but since the #14932 temporary fix went in I don't think this is a 2.17 blocker anymore and I want to stay focused. So I'm moving it to the 2.18 list. As always, disagree and re-tag if you feel strongly, I'm okay with that

**Comment by david-allison** [comment]
* Likely fixed by https://github.com/ankidroid/Anki-Android/issues/15679, please reopen if I'm incorrect
